tests/test_pip_repositories.py

```diff
@@ -0,0 +1,172 @@
+import base64
+import os
+import tarfile
+
+from io import BytesIO
+from pathlib import Path
+from typing import Optional, Tuple
+from urllib.parse import urlparse
+
+import pytest
+import requests
+import requests_mock
+
+from conda_lock.conda_lock import DEFAULT_LOCKFILE_NAME, run_lock
+from conda_lock.lockfile import parse_conda_lock_file
+from tests.test_conda_lock import clone_test_dir
+
+
+_PRIVATE_REPO_USERNAME = "secret-user"
+_PRIVATE_REPO_PASSWORD = "secret-password"
+
+_PRIVATE_REPO_ROOT = """<!DOCTYPE html>
+<html>
+  <body>
+    <a href="/api/pypi/simple/fake-private-package/">fake-private-package</a>
+  </body>
+</html>
+"""
+
+_PRIVATE_REPO_PACKAGE = """<!DOCTYPE html>
+<html>
+  <body>
+    <a href="/files/fake-private-package-1.0.0.tar.gz">fake-private-package-1.0.0.tar.gz</a>
+  </body>
+</html>
+"""
+
+_PRIVATE_PACKAGE_SDIST_PATH = (
+    Path(__file__).parent / "test-pip-repositories" / "fake-private-package-1.0.0"
+)
+
+
+@pytest.fixture(scope="module")
+def private_package_tar():
+    tar_path = _PRIVATE_PACKAGE_SDIST_PATH.parent / "fake-private-package-1.0.0.tar.gz"
+    with tarfile.open(tar_path, "w:gz") as tar:
+        tar.add(
+            _PRIVATE_PACKAGE_SDIST_PATH,
+            arcname=os.path.basename(_PRIVATE_PACKAGE_SDIST_PATH),
+        )
+    try:
+        yield tar_path
+    finally:
+        os.remove(tar_path)
+
+
+@pytest.fixture(autouse=True)
+def mock_private_pypi(private_package_tar: Path):
```

Comment:
This is really impressive. On the other hand, I'm honestly a bit concerned about the maintenance overhead for this. It'd be really annoying if we have to modify this in the future.

I wonder if you could publish this as a separate package? Then hopefully others will start using it so that it can be community-maintained. And anyways, this is really great work.

That said, I'm very grateful to accept this as-is.

Reply:
I think to publish this as a separate package, it would likely need to implement a broader set of PyPi behaviour - at which point it comes close to an actual implementation PyPi implementation.

Some possible alternatives:

- A more robust and complete solution might be to use [the pypiserver docker image](https://hub.docker.com/r/pypiserver/pypiserver), and bring this up as a backing service whilst running the tests. This is pretty simple to configure with tools like [pytest-docker](https://pypi.org/project/pytest-docker/) and means any PyPi interaction is covered realistically. However it also means the test pipeline needs to be run on a host with docker available, and slows things down as the image needs to be pulled in CI.
- An equivalent but probably simpler implementation could probably be achieved by borrowing the test framework from a popular Python web framework like FastAPI. That way the request parsing and response construction could be offloaded to a reliable framework.

I do think there is unlikely to be much need to modify this in future, as the PyPI interface changes very rarely. I imagine its more likely that it might need extending if we wanted `conda-lock` to interact with PyPI in new ways - that also seems unlikely to me but I imagine you'd have a better idea.

Reply:
I don't have much sense of how often PyPI changes their interface, but I'm happy to use this for now. In case it breaks in the future we can think about your suggested alternatives.